### PR TITLE
[ember] Add declarations for strict-mode template entities

### PR DESCRIPTION
**Pull request by dfreeman**
<details>
<summary>Checklists</summary>

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: (details below)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

</details>

This change introduces declarations for importable entities for use in Ember's templating system as defined in [the Handlebars Strict Mode RFC](https://github.com/emberjs/rfcs/blob/master/text/0496-handlebars-strict-mode.md#handlebars-strict-mode).

That RFC introduces two packages that previously didn't exist specifically for hosting these entities (`@ember/helper` and `@ember/modifier`) as well as adding exports to `@ember/routing` and `@ember/component`.

@chriskrycho per our recent Discord conversation I went with augmentable interfaces extending from `Opaque` for the types of these items so that we can merge template-specific type info from e.g. Glint. If you've had any ideas in the meantime for other ways to approach this, though, I'm happy to update.

### CODE REVIEW

File: types/ember__component/index.d.ts [hunk 1]
```diff
@@ -8,10 +8,11 @@
 
 /// <reference types="jquery" />
 
-import CoreView from "@ember/component/-private/core-view";
-import ClassNamesSupport from "@ember/component/-private/class-names-support";
-import ViewMixin from "@ember/component/-private/view-mixin";
+import CoreView from '@ember/component/-private/core-view';
```
Comment (dfreeman) [inline types/ember__component/index.d.ts:11]: Looks like Prettier got itself involved in this file when I saved without me noticing 🙃

File: types/ember__helper/ember__helper-tests.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { array, concat, fn, get, hash } from '@ember/helper';
+
+array; // $ExpectType ArrayHelper
+concat; // $ExpectType ConcatHelper
+fn; // $ExpectType FnHelper
+get; // $ExpectType GetHelper
+hash; // $ExpectType HashHelper
```
Comment (dfreeman) [inline types/ember__helper/ember__helper-tests.ts:7]: Given the only thing you can do with these values is reference them in a template, the most I think we can do in testing is assert that they're importable and have the expected type label.

File: types/ember__component/index.d.ts (hunk 1 above)
Comment (chriskrycho) [reply types/ember__component/index.d.ts:11]: THE WORST.

(We should do another Prettier pass on all of these at some point. Or: not, and get them into Ember itself. :joy:)

File: types/ember__component/index.d.ts [hunk 3]
```diff
@@ -8,10 +8,11 @@
 
 /// <reference types="jquery" />
```
Comment (chriskrycho) [inline types/ember__component/index.d.ts:9]: Huh, looks like I missed this doing the v4 cleanup. I'll do a follow-on PR for it.

**Review by chriskrycho — APPROVED** [review]
Thanks, @dfreeman! 👍🏼

File: types/ember__component/index.d.ts [hunk 4]
```diff
@@ -106,10 +104,28 @@ export default class Component extends CoreView.extend(
...
+ * The `Input` component lets you create an HTML `<input>` element.
+ *
+ * @see https://api.emberjs.com/ember/4.1/classes/Ember.Templates.components/methods/Input?anchor=Input
+ */
+export const Input: Input;
+
+/**
+ * The `Textarea` component inserts a new instance of `<textarea>` tag into the template.
+ *
+ * @see https://api.emberjs.com/ember/4.1/classes/Ember.Templates.components/methods/Textarea?anchor=Textarea
```
Comment (amcasey) [inline types/ember__component/index.d.ts:126]: The URL says 4.1 but the file header says 4.0 - should it be updated?

**Review by amcasey — APPROVED** [review]
One question about the version number, but LGTM

File: types/ember__component/index.d.ts (hunk 4 above)
Comment (chriskrycho) [reply types/ember__component/index.d.ts:126]: Good question. That’s the latest version of Ember docs at present but the API is available from 4.0 forward. I think docs are only available later for uninteresting reasons.

Comment (amcasey) [reply types/ember__component/index.d.ts:126]: If the header reflects the API version, that's good enough for me. 😄

**Comment by dfreeman** [comment]
Ready to merge

**Pull request merged**